src/aws/flb_aws_util.c

```diff
@@ -927,3 +932,73 @@ flb_sds_t flb_get_s3_key(const char *format, time_t time, const char *tag,
         }
         return NULL;
 }
+
+/*
+ * Adjust the time_key output format according to time_key_format.
+ */
+int create_time_format_string(char **out_buff_ptr, size_t *out_size,
+                              const char *time_key_format, struct flb_time *tms)
+{
+    char millisecond_string[FLB_AWS_MILLISECOND_STRING_LENGTH+1];
+    char nanosecond_string[FLB_AWS_NANOSECOND_STRING_LENGTH+1];
+    char *tmp_buff_time_str;
+    char *mem_for_time_key_format;
+    char *tmp_out_buff;
+    size_t tmp_out_size;
+    size_t max_length_for_mem;
+    struct tm time_stamp;
+
+    /* guess the maximum length needed for out_buff, the worst scenario will be only %c in time_key_format,
+    and then %c will be tansfer to 24 chars by strftime, so update length to 12 times to avoid memory overflow. */
+    max_length_for_mem = 12*strlen(time_key_format);
+    *out_buff_ptr = (char *)flb_calloc(max_length_for_mem, sizeof(char));
```

Comment:
Please allocate the final memory, the memory you return, at the end of the function once its all ready.

Reply:
Basically, only alloc the return memory once you know the time formatting has succeeded, this will be most efficient. This memory isn't used till end of function.